drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,913 @@
+ * 160 Mhz PLL: 1
+ * When using PLL root clock, we can use either setting, when using the 32Mhz Oscillator with a
+ * uninitialized PLL, only FCLK will be available.
+ */
+static void clock_control_bl60x_uart_set_clock(bool enable, uint32_t clock, uint32_t divider)
+{
+	uint32_t tmp;
+
+	if (divider > 0x7) {
+		divider = 0x7;
+	}
+	if (clock > 1) {
+		clock = 1;
+	}
+	/* disable uart clock */
+	clock_control_bl60x_uart_set_clock_enable(false);
+
+	tmp = sys_read32(GLB_BASE + GLB_CLK_CFG2_OFFSET);
+	tmp = (tmp & GLB_UART_CLK_DIV_UMSK) | (divider << GLB_UART_CLK_DIV_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CLK_CFG2_OFFSET);
+
+	tmp = sys_read32(HBN_BASE + HBN_GLB_OFFSET);
+	tmp = (tmp & HBN_UART_CLK_SEL_UMSK) | (clock << HBN_UART_CLK_SEL_POS);
+	sys_write32(tmp, HBN_BASE + HBN_GLB_OFFSET);
+
+	clock_control_bl60x_uart_set_clock_enable(enable);
+}
+
+/* Simple function to enable all peripherals for now */
+static void clock_control_bl60x_peripheral_clock_init(void)
+{
+	uint32_t regval = sys_read32(GLB_BASE + GLB_CGEN_CFG1_OFFSET);
+
+	/* enable ADC clock routing */
+	regval |= (1 << 2);
+	/* enable UART0 clock routing */
+	regval |= (1 << 16);
+	/* enable I2C0 clock routing */
+	regval |= (1 << 19);
+
+	sys_write32(regval, GLB_BASE + GLB_CGEN_CFG1_OFFSET);
+
+	clock_control_bl60x_uart_set_clock(1, 0, 0);
+}
+
+static int clock_control_bl60x_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_bl60x_data *data = dev->data;
+	int ret = -EINVAL;
+	uint32_t key;
+	enum bl60x_clkid oldroot;
+
+	key = irq_lock();
+
+	if ((enum bl60x_clkid)sys == bl60x_clkid_clk_crystal) {
+		if (data->crystal_enabled) {
+			ret = 0;
+		} else {
+			data->crystal_enabled = true;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->crystal_enabled = false;
+			}
+		}
+	} else if ((enum bl60x_clkid)sys == bl60x_clkid_clk_pll) {
+		if (data->pll_enabled) {
+			ret = 0;
+		} else {
+			data->pll_enabled = true;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->pll_enabled = false;
+			}
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_RC32M) {
+		if (data->root.source == bl60x_clkid_clk_rc32m) {
+			ret = 0;
+		} else {
+			/* Cannot fail to set root to rc32m */
+			data->root.source = bl60x_clkid_clk_rc32m;
+			ret = clock_control_bl60x_update_root(dev);
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_CRYSTAL) {
+		if (data->root.source == bl60x_clkid_clk_crystal) {
+			ret = 0;
+		} else {
+			oldroot = data->root.source;
+			data->root.source = bl60x_clkid_clk_crystal;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->root.source = oldroot;
+			}
+		}
+	} else if ((int)sys == BFLB_FORCE_ROOT_PLL) {
+		if (data->root.source == bl60x_clkid_clk_pll) {
+			ret = 0;
+		} else {
+			oldroot = data->root.source;
+			data->root.source = bl60x_clkid_clk_pll;
+			ret = clock_control_bl60x_update_root(dev);
+			if (ret < 0) {
+				data->root.source = oldroot;
+			}
+		}
+	}
+
```

Comment:
nits: I think it's better to set `ret = -EINVAL;` in the else case here and `clock_control_bl60x_off()`

Reply:
The default negative logic will stay.